### PR TITLE
unity.cpp sync

### PR DESCRIPTION
**Pull request by cndn**
Sync unity.cpp from internal repo

Major updates:
(1) Sync with ggml official repo in December
(2) Support both SeamlessM4T (X2T parts) and NLLB loading
(3) Add word-level confidence score and LID for multilingual models 
(4) Memory allocation optimization 

TODO: Update README

### CODE REVIEW

File: ggml/ggml_convert.py [hunk 1]
```diff
@@ -6,54 +6,330 @@
...
+    vocab = read_vocab(tokenizer)
+
+    return model, hparams, vocab
+
+
+def convert_bitext_model(
+    model_name: str,
+    hparams: Optional[Dict[str, Any]] = None,
+):
+    from mt import load_mt_model, load_vocab  #, test_mt
```
Comment (antoine-tran) [inline ggml/ggml_convert.py:259]: I think it is fine to upstream mt.py here . Its general purpose is to support backward compatibilty for fairseq-based text translation models that use standard transformer. 

Either that, or we skip this function (and trim the permissible module type above)

File: ggml/README.md [hunk 2]
```diff
@@ -34,7 +31,7 @@ In the console, enter the path of local waveform file and target language, separ
 Converted ggml models could be downloaded from 
 |SeamlessM4T_large | SeamlessM4T_medium | 
 |-------- | -------- | 
-| [model](https://dl.fbaipublicfiles.com/seamless/models/seamlessM4T_large.ggml) | [model](https://dl.fbaipublicfiles.com/seamless/models/seamlessM4T_medium.ggml) |  
+| [model](dl.fbaipublicfiles.com/seamless/models/seamlessM4T_large.ggml) | [model](dl.fbaipublicfiles.com/seamless/models/seamlessM4T_medium.ggml) |
```
Comment (antoine-tran) [inline ggml/README.md:34]: nit: Could we upload some example NLLB models (like nllb-200_dense_distill_600m.ggml) too ?

**Review by antoine-tran — COMMENTED** [review]
the PR looks good to me. Leaving some minor comments about the decision of adding mt.py vs. trimming mt-related conversion functions, and more advocacies for nllb ggml checkpoints. Thanks !

File: ggml/examples/unity/unity.cpp [hunk 3]
```diff
@@ -179,46 +150,47 @@ int main(int argc, char ** argv) {
...
-        ggml_set_no_alloc(model.ctx, true);
+        // Load audio input
         GGML_ASSERT(info.samplerate == 16000);
         GGML_ASSERT(info.channels == 1);
-        // Truncate audio input. Ideally we should chunk it, but this will prevent most obvious OOM.
-        int n_frames = std::min(info.samplerate * params.max_audio_s, (int)info.frames);
-        ggml_tensor* seqs = ggml_new_tensor_2d(model.ctx, GGML_TYPE_F32, n_frames, info.channels);
-        ggml_allocr_alloc(fwd_alloc, seqs);
+        // stop at 30s. Ideally we should chunk input audio, but this will prevent most obvious OOM.
+        int n_frames = std::min(info.samplerate * 30, (int)info.frames);
```
Comment (gwenzek) [inline ggml/examples/unity/unity.cpp:157]: max_audio_s was used here to allow longer samples. Agree this is a hack, waiting for a cleaner solution, but we don't have fully decided how to implement this.

**Review by gwenzek — APPROVED** [review]
Accepting, since those changes have been tested internally, be careful with the "audio_max_s" conflict though.

Can you also document how many GGML modifications we still have after the upgrade ?

**Pull request merged**